x/programs/cmd/simulator/cmd/program.go

```diff
@@ -125,21 +118,16 @@ func programExecuteFunc(
 		FunctionName: function,
 		Params:       bytes,
 	}
-	programOutput, err := rt.CallProgram(ctx, callInfo)
-	output := [][]byte{programOutput}
+	result, err := rt.CallProgram(ctx, callInfo)
+	remainingFuel := callInfo.RemainingFuel()
 	if err != nil {
-		response := multilineOutput(output)
-		return nil, 0, fmt.Errorf("program execution failed: %s, err: %w", response, err)
+		if code, ok := runtime.ExtractErrorCode(err); ok {
+			return runtime.Err[runtime.RawBytes, runtime.ErrorCode](code), remainingFuel, nil
+		}
+		return runtime.Err[runtime.RawBytes, runtime.ErrorCode](runtime.ExecutionFailure), remainingFuel, err
```

Comment:
We used to waste all the allocated gas if it was returning an error, now the rest is remaining.